src/test/java/school/hei/haapi/integration/CourseIT.java

```diff
@@ -0,0 +1,352 @@
+    course.setName("Advanced OOP");
+    return course;
+  }
+
+  static Course course3() {
+    Course course = new Course();
+    course.setId("course3_id");
+    course.setCode("WEB1");
+    course.setCredits(4);
+    course.setTotalHours(16);
+    course.setMainTeacher(teacher2());
+    course.setName("Web Interface");
+    return course;
+  }
+
+
+  static Course course4() {
+    Course course = new Course();
+    course.setId(COURSE4_ID);
+    course.setCode("WEB2");
+    course.setCredits(6);
+    course.setTotalHours(18);
+    course.setMainTeacher(teacher3());
+    course.setName("Web Application");
+    return course;
+  }
+
+  static Course course5() {
+    Course course = new Course();
+    course.setId(COURSE5_ID);
+    course.setCode("MGT1");
+    course.setCredits(5);
+    course.setTotalHours(12);
+    course.setMainTeacher(teacher4());
+    course.setName("Collaborative work");
+    return course;
+  }
+
+  static CrupdateCourse crupdatedCourse1() {
+    CrupdateCourse crupdatedCourse = new CrupdateCourse();
+    crupdatedCourse.setId(COURSE5_ID);
+    crupdatedCourse.setCode("MGT1");
+    crupdatedCourse.setName("Collaborative work");
+    crupdatedCourse.setCredits(5);
+    crupdatedCourse.setTotalHours(12);
+    crupdatedCourse.setMainTeacherId(TEACHER4_ID);
+    return crupdatedCourse;
+  }
+
+  static CrupdateCourse crupdatedCourse2() {
+    CrupdateCourse crupdatedCourse = new CrupdateCourse();
+    crupdatedCourse.setCode("MGT1");
+    crupdatedCourse.setName("Collaborative work like GWSP");
+    crupdatedCourse.setCredits(12);
+    crupdatedCourse.setTotalHours(5);
+    crupdatedCourse.setMainTeacherId(TEACHER4_ID);
+    return crupdatedCourse;
+  }
+
+  static UpdateStudentCourse updateStudentCourse() {
+    UpdateStudentCourse updated = new UpdateStudentCourse();
+    updated.setCourseId(COURSE3_ID);
+    updated.status(LINKED);
+    return updated;
+  }
+
+  @BeforeEach
+  void setUp() {
+    setUpCognito(cognitoComponentMock);
+  }
+
+  private boolean isBefore(String a, String b) {
+    return a.compareTo(b) < 0;
+  }
+
+  private boolean isBefore(int a, int b) {
+    return a < b;
+  }
+
+  @Test
+  void student_read_ok() throws ApiException {
+    ApiClient student1Client = anApiClient(STUDENT1_TOKEN);
+    TeachingApi api = new TeachingApi(student1Client);
+
+    List<Course> actual = api.getCourses(null, null, null, null, null,
+        null, null, null, null);
+
+    assertEquals(5, actual.size());
+    assertTrue(actual.contains(course1()));
+  }
+
+  @Test
+  void teacher_read_ok() throws ApiException {
+    ApiClient teacher1Client = anApiClient(TEACHER1_TOKEN);
+    TeachingApi api = new TeachingApi(teacher1Client);
+
+    List<Course> actual = api.getCourses(null, null, null, null, null,
+        null, null, null, null);
+
+    assertEquals(5, actual.size());
+    assertTrue(actual.contains(course2()));
+  }
+
+  @Test
+  void user_read_by_code() throws ApiException {
```

Comment:
No.

Don't separate the `filter by criteria` test.

Put it inside an unique test with different scenarios.

Also, don't forget to insert a course with some null values inside the database. For example, only the code, the name and the main teacher ID are given and others null. Use test for this.

Reply:
But for the course table, we add a not null constraint to every column so should i drop them all ?